src/crimson/os/seastore/randomblock_manager.cc

```diff
@@ -0,0 +1,667 @@
+		bool out = false;
+		for (auto b : allocated_blocks) {
+		  if (b.alloc_blk_ids.intersects(block_id, 1)) {
+		    out = true;
+		    break;
+		  }
+		}
+		if (out) {
+		  continue;
+		}
+		if (b_block.is_allocated(i)) {
+		  continue;
+		}
+		logger().debug("find_free_list: allocated block no {} i {}",
+			       convert_bitmap_block_no_to_block_id(i, addr), i);
+		if (allocated != 0 && alloc_extent.range_end() != block_id) {
+		  /*
+		   * if not continous block, just restart to find continuous blocks at the next block.
+		   * in-memory allocator can handle this efficiently.
+		   */
+		  allocated = 0;
+		  alloc_extent.clear(); // a range of block allocation
+		  logger().debug("find_free_list: rety to find continuous blocks");
+		  continue;
+		}
+		allocated += 1;
+		alloc_extent.insert(block_id);
+	      }
+	      addr += super.block_size;
+	      logger().debug("find_free_list: allocated: {} alloc_extent {}", allocated, alloc_extent);
+	      if (((uint64_t)size)/super.block_size == allocated) {
+		return find_block_ertr::make_ready_future<bool>(true); 
+	      } else if (addr >= super.start_data_area) {
+		alloc_extent.clear();
+		return find_block_ertr::make_ready_future<bool>(true); 
+	      }
+	      return find_block_ertr::make_ready_future<bool>(false); 
+	      });
+	}).safe_then([&allocated, &alloc_extent, &t, size, this] () {
+	  logger().debug(" allocated: {} size {} ", allocated * super.block_size, size);
+	  if (allocated * super.block_size < size) {
+	    alloc_extent.clear();
+	  }
+	  return find_block_ret(
+	      find_block_ertr::ready_future_marker{},
+	      alloc_extent);
+	  }).handle_error(
+	      find_block_ertr::pass_further{},
+	      crimson::ct_error::assert_all{
+		"Invalid error in RandomBlockManager::find_free_block"
+	      }
+	    );
+    });
+}
+
+/* TODO : block allocator */
+RandomBlockManager::allocate_ertr::future<> RandomBlockManager::alloc_extent(Transaction &t, size_t size)
+{
+  
+  /* 
+   * 1. find free blocks using block allocator
+   * 2. add free blocks to transaction (free block is reserved state, but not stored)
+   * 3. link free blocks to onode
+   * Due to in-memory block allocator is the next work to do,
+   * just read the block bitmap directly to find free blocks here.
+   *
+   */
+  return find_free_block(t, size
+      ).safe_then([this, &t, size] (auto alloc_extent) mutable -> allocate_ertr::future<> {
+	logger().debug("after find_free_block: allocated {}", alloc_extent);
+	if (!alloc_extent.empty()) {
+	  // add alloc info to delta
+	  rbm_alloc_delta_t alloc_info {
+	    extent_types_t::RBM_ALLOC_INFO,
+	    alloc_extent,
+	    rbm_alloc_delta_t::op_types_t::SET 
+	  };
+	  t.add_rbm_allocated_blocks(alloc_info);
+	} else {
+	  return crimson::ct_error::enospc::make();
+	}
+	return allocate_ertr::now();
+	}
+      ).handle_error(
+	allocate_ertr::pass_further{},
+	crimson::ct_error::assert_all{
+	  "Invalid error find_free_block in RandomBlockManager::alloc_extent"
+	}
+	);
```

Comment:
I don't see any in-memory state here that remembers pending allocations between alloc/free_extent and complete_allocation.  Is this still todo?

Reply:
Yes, it is todo. I'll address what you pointed out next PR.